storagenode/piecestore/endpoint.go

```diff
@@ -166,6 +168,42 @@ func (endpoint *Endpoint) Delete(ctx context.Context, delete *pb.PieceDeleteRequ
 	return &pb.PieceDeleteResponse{}, nil
 }
 
+// DeletePiece handles deleting a piece on piece store requested by satellite.
+//
+// It doesn't return an error if the piece isn't found by any reason.
+func (endpoint *Endpoint) DeletePiece(
+	ctx context.Context, req *pb.DeletePieceRequest,
+) (_ *pb.DeletePieceResponse, err error) {
+	defer mon.Task()(&ctx, req.PieceId.String())(&err)
+
+	peer, err := identity.PeerIdentityFromContext(ctx)
+	if err != nil {
+		return nil, rpcstatus.Error(rpcstatus.Unauthenticated, Error.Wrap(err).Error())
+	}
+
+	err = endpoint.trust.VerifySatelliteID(ctx, peer.ID)
+	if err != nil {
+		return nil, rpcstatus.Error(rpcstatus.PermissionDenied,
+			Error.New("%s", "delete piece called with untrusted ID").Error(),
+		)
+	}
```

Comment:
I am not really sure if we need this part. The main reason for not trusting a satellite is to avoid getting data from the uplinks of this satellite.

If the satellite (or fake peer) was not trusted initially then the storage node would not have any pieces from this peer. So it should be safe to continue with the `Delete` operations below, which would return an error like NotFound.

On the other hand, if the satellite was initially trusted, the storage node got some pieces, and then the satellite was not trusted anymore, then I don't see a reason to not process delete requests from this satellite, even it is not trusted anymore. Why should the storage node continue keeping these pieces?

@jtolds @zeebo @egonelbre thoughts?

Reply:
> which would return an error like NotFound.

NOTE that not found piece don't return an error it returns OK because the store Delete function (called in https://github.com/storj/storj/pull/3590/files/9355da53c6768c06db450446bad2a8163daf3474#diff-c8abc956f27b79755a16b8da2c392cbbR191) doesn't return an error if the piece isn't found.

> On the other hand, if the satellite was initially trusted, the storage node got some pieces, and then the satellite was not trusted anymore, then I don't see a reason to not process delete requests from this satellite, even it is not trusted anymore. Why should the storage node continue keeping these pieces?

Can a satellite fake a piece ID generated by another satellite?

Reply:
> Can a satellite fake a piece ID generated by another satellite?

No, no peer can fake their peer ID. The identity returned by `identity.PeerIdentityFromContext(ctx)` is always authentic.

Reply:
Then I don't have any other concern for removing the check.

Reply:
To me it makes sense to disallow all operations from a satellite you don't trust. Maybe a satellite becomes publicly malicious or hacked and wants to target some user's data? So maybe a storagenode deciding to delete pieces from an untrusted satellite should probably be the SNO's decision, not the untrusted satellite's decision.

Reply:
> No, no peer can fake their peer ID. The identity returned by `identity.PeerIdentityFromContext(ctx)` is always authentic.

Yeah, but he was asking about piece ID, not peer ID. It's entirely possible that a bad satellite could get hold of valid piece IDs belonging to other satellites. We don't treat piece IDs as highly sensitive information.

Reply:
@thepaul you're right I asked for it but with @kaloyan-raev made me think that even though a satellite could send a valid pieceID corresponding to another satellite the `endpoint.store.Delete` takes a peer ID as a parameter and if I'm not wrong it acts as a namespace so it should not be able to delete pieces corresponding to another satellite.

Reply:
> To me it makes sense to disallow all operations from a satellite you don't trust

It makes more sense to mee too.

> Maybe a satellite becomes publicly malicious or hacked and wants to target some user's data? So maybe a storagenode deciding to delete pieces from an untrusted satellite should probably be the SNO's decision, not the untrusted satellite's decision.

This is a good case, however, I'm thinking now that we should provide to SNO's an easy way (e.g. a command) to delete pieces from untrusted satellites.

In the example that you exposed, the SNO could decide not to delete the pieces from an untrusted satellite recently removed because it was announced that the satellite got hacked and the satellite operators indicated to so until they can ensure that the satellite is secured again. Deleting all the pieces would wipe all the data of the network controlled by that satellite while holding them doesn't.

On the other hand, an SNO could decide to untrust a satellite because it's not paying or whatever other reason and wants to release the disk space occupied by its data.

Reply:
OK. Let's leave this check. I'll add the question to Arch Review meeting on how storage nodes should treat untrusted satellites in general.

Reply:
I resolve this conversation and leaving the check for untrusted satellite requests as we commented in the arch meeting.